src/modules/keyboardmanager/common/KeyDelay.cpp

```diff
@@ -0,0 +1,167 @@
+#include "pch.h"
+#include "KeyDelay.h"
+
+#define LONG_PRESS_DELAY_MILLIS 900
+#define ON_HOLD_WAIT_TIMEOUT_MILLIS 50
+
+KeyDelay::~KeyDelay()
+{
+    _quit = true;
+    _cv.notify_all();
+    _delayThread.join();
+}
+
+void KeyDelay::KeyEvent(LowlevelKeyboardEvent* ev)
+{
+    std::lock_guard guard(_queueMutex);
+    _queue.push({ ev->lParam->time, ev->wParam });
+    _cv.notify_all();
+}
+
+KeyTimedEvent KeyDelay::NextEvent()
+{
+    auto ev = _queue.front();
+    _queue.pop();
+    return ev;
+}
+
+bool KeyDelay::CheckIfMillisHaveElapsed(DWORD first, DWORD last, DWORD duration)
+{
+    if (first < last && first <= first + duration) 
+    {
+        return first + duration < last;
+    }
+    else
+    {
+        first += ULONG_MAX / 2;
+        last += ULONG_MAX / 2;
+        return first + duration < last;
+    }
+}
+
+bool KeyDelay::HasNextEvent()
+{
+    return !_queue.empty();
+}
+
+bool KeyDelay::HandleRelease()
+{
+    while (HasNextEvent())
+    {
+        auto ev = NextEvent();
+        switch (ev.message)
+        {
+        case WM_KEYDOWN:
+        case WM_SYSKEYDOWN:
+            _state = KeyDelayState::ON_HOLD;
+            _initialHoldKeyDown = ev.time;
+            return false;
+        case WM_KEYUP:
+        case WM_SYSKEYUP:
+            break;
+        }
+    }
+
+    return true;
+}
+
+bool KeyDelay::HandleOnHold(std::unique_lock<std::mutex>& cvLock)
+{
+    while (HasNextEvent())
+    {
+        auto ev = NextEvent();
+        switch (ev.message)
+        {
+        case WM_KEYDOWN:
+        case WM_SYSKEYDOWN:
+            break;
+        case WM_KEYUP:
+        case WM_SYSKEYUP:
+            if (CheckIfMillisHaveElapsed(_initialHoldKeyDown, ev.time, LONG_PRESS_DELAY_MILLIS))
+            {
+                if (_onLongPressDetected != nullptr)
+                {
+                    _onLongPressDetected(_key);
+                }
+                if (_onLongPressReleased != nullptr)
+                {
+                    _onLongPressReleased(_key);
+                }
+            }

```

Comment:
Just to clarify, I'm guessing this is needed for those border-line cases where the key is released right after the threshold but the last keydown was just below the threshold?

Reply:
Yes, also since the events have the millisecond the key press happened it seemed appropriate to use it to be a little bit more exact.